File: 2025-02-February-GFG-POTD/February 04 - Diameter of a Binary Tree/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;
    Node(int val) {
        data = val;
        left = right = nullptr;
    }
};

class Solution {
public:
    pair<int, int> f(Node* root) {
        if (!root) return {0, 0};
        
        auto left = f(root->left);
        auto right = f(root->right);
        
        int height = 1 + max(left.first, right.first);
        int diameter = max({left.second, right.second, left.first + right.first});
        
        return {height, diameter};
    }

    int diameter(Node* root) {
        return f(root).second;
    }
};

// Function to build tree from input string
Node* buildTree(string str) {
    if (str.empty() || str[0] == 'N') return nullptr;
    
    vector<string> ip;
    stringstream ss(str);
    string word;
    while (ss >> word) ip.push_back(word);

    Node* root = new Node(stoi(ip[0]));
    queue<Node*> q;
    q.push(root);
    
    int i = 1;
    while (!q.empty() && i < ip.size()) {
        Node* currNode = q.front();
        q.pop();

        string leftVal = ip[i++];
        if (leftVal != "N") {
            currNode->left = new Node(stoi(leftVal));
            q.push(currNode->left);
        }

        if (i >= ip.size()) break;
        string rightVal = ip[i++];
        if (rightVal != "N") {
            currNode->right = new Node(stoi(rightVal));
            q.push(currNode->right);
        }
    }

    return root;
}

// Driver code
int main() {
    int t;
    cin >> t;
    cin.ignore();

    while (t--) {
        string s;
        getline(cin, s);
        Node* root = buildTree(s);
        Solution sol;
        cout << sol.diameter(root) << endl;
    }

    return 0;
}

File: 2025-02-February-GFG-POTD/February 04 - Diameter of a Binary Tree/Solution.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class Solution:
    def f(self, root):
        if not root:
            return (0, 0)
        
        left_height, left_diameter = self.f(root.left)
        right_height, right_diameter = self.f(root.right)

        height = 1 + max(left_height, right_height)
        diameter = max(left_diameter, right_diameter, left_height + right_height)

        return (height, diameter)

    def diameter(self, root):
        return self.f(root)[1]

# Function to build tree from input string
def buildTree(s):
    if not s or s[0] == 'N':
        return None

    ip = s.split()
    root = Node(int(ip[0]))
    q = [root]
    
    i = 1
    while q and i < len(ip):
        currNode = q.pop(0)

        leftVal = ip[i]
        if leftVal != "N":
            currNode.left = Node(int(leftVal))
            q.append(currNode.left)

        i += 1
        if i >= len(ip):
            break

        rightVal = ip[i]
        if rightVal != "N":
            currNode.right = Node(int(rightVal))
            q.append(currNode.right)

        i += 1

    return root

# Driver code
if __name__ == "__main__":
    t = int(input().strip())

    for _ in range(t):
        s = input().strip()
        root = buildTree(s)
        sol = Solution()
        print(sol.diameter(root))
